File: src/services/fuelEngine.ts
import type { VehicleConfig } from '../domain/models/telemetry';

export const DEFAULT_VEHICLE_CONFIG: VehicleConfig = {
  vehicleName: 'Toyota Lanka Cruiser (2.0L)',
  vehicleType: 'SEDAN',
  fuelType: 'GASOLINE',
  engineSizeLiters: 2.0,
  idleConsumptionRateLph: 1.2, // 1.2 Liters/hour idling
  fuelPricePerLiter: 370.0,    // 370 LKR per liter (Sri Lankan Octane 92 / Diesel)
};

export class FuelEngine {
  private config: VehicleConfig;

  constructor(config: VehicleConfig = DEFAULT_VEHICLE_CONFIG) {
    this.config = config;
  }

  public updateConfig(newConfig: Partial<VehicleConfig>): VehicleConfig {
    this.config = { ...this.config, ...newConfig };
    return this.config;
  }

  public getConfig(): VehicleConfig {
    return { ...this.config };
  }

  /**
   * Calculate fuel wasted and monetary cost in LKR (Sri Lankan Rupees) during idle stationary state.
   */
  public calculateIdleWaste(idleDurationSeconds: number): {
    idleLiters: number;
    idleCost: number; // LKR
    costPerMinute: number; // LKR / min
    co2EmissionsKg: number;
  } {
    const idleHours = idleDurationSeconds / 3600;
    const idleLiters = idleHours * this.config.idleConsumptionRateLph;
    const idleCost = idleLiters * this.config.fuelPricePerLiter;
    
    // Cost per minute while stuck idling in LKR
    const costPerMinute = (this.config.idleConsumptionRateLph / 60) * this.config.fuelPricePerLiter;

    // Approx CO2 emissions per Liter of gasoline = ~2.31 kg CO2
    const co2EmissionsKg = idleLiters * 2.31;

    return {
      idleLiters: parseFloat(idleLiters.toFixed(3)),
      idleCost: parseFloat(idleCost.toFixed(2)),
      costPerMinute: parseFloat(costPerMinute.toFixed(2)),
      co2EmissionsKg: parseFloat(co2EmissionsKg.toFixed(2)),
    };
  }

  /**
   * Calculate overall trip fuel cost breakdown in LKR.
   */
  public calculateTripFuelMetrics(
    distanceKm: number,
    idleDurationSeconds: number,
    avgSpeedKmH: number
  ): {
    totalFuelLiters: number;
    idleFuelLiters: number;
    totalCost: number; // LKR
    idleCost: number; // LKR
    costSavedCruising: number; // LKR
    cruisingEfficiencyLPer100Km: number;
  } {
    const { idleLiters, idleCost } = this.calculateIdleWaste(idleDurationSeconds);

    // Cruising efficiency model (approx 6.5 - 9.0 L/100km depending on engine size)
    const baseLPer100Km = 5.5 + this.config.engineSizeLiters * 1.2;
    
    // Speed efficiency curve multiplier (optimal speed ~60-80 km/h)
    let speedPenalty = 1.0;
    if (avgSpeedKmH > 0 && avgSpeedKmH < 30) {
      speedPenalty = 1.35; // Stop-and-go traffic jam penalty
    } else if (avgSpeedKmH > 100) {
      speedPenalty = 1.25; // High speed drag penalty
    }

    const cruisingLPer100Km = baseLPer100Km * speedPenalty;
    const cruisingLiters = (distanceKm / 100) * cruisingLPer100Km;
    const totalFuelLiters = cruisingLiters + idleLiters;

    const cruisingCost = cruisingLiters * this.config.fuelPricePerLiter;
    const totalCost = cruisingCost + idleCost;

    // Hypothetical savings if 70% of idle traffic jam time was smooth cruising
    const hypotheticalSavedIdleLiters = idleLiters * 0.7;
    const costSavedCruising = hypotheticalSavedIdleLiters * this.config.fuelPricePerLiter;

    return {
      totalFuelLiters: parseFloat(totalFuelLiters.toFixed(2)),
      idleFuelLiters: idleLiters,
      totalCost: parseFloat(totalCost.toFixed(2)),
      idleCost,
      costSavedCruising: parseFloat(costSavedCruising.toFixed(2)),
      cruisingEfficiencyLPer100Km: parseFloat(cruisingLPer100Km.toFixed(1)),
    };
  }
}

File: smart_trip_ai_mobile/lib/services/fuel_engine.dart
import '../domain/models/telemetry.dart';

final VehicleConfig defaultVehicleConfig = VehicleConfig(
  vehicleName: 'Toyota Lanka Cruiser (2.0L)',
  vehicleType: VehicleType.sedan,
  fuelType: FuelType.gasoline,
  engineSizeLiters: 2.0,
  idleConsumptionRateLph: 1.2, // 1.2 Liters/hour idling
  fuelPricePerLiter: 370.0,    // 370 LKR per liter (Sri Lankan Octane 92 / Diesel)
);

class IdleWasteResult {
  final double idleLiters;
  final double idleCost;
  final double costPerMinute;
  final double co2EmissionsKg;

  IdleWasteResult({
    required this.idleLiters,
    required this.idleCost,
    required this.costPerMinute,
    required this.co2EmissionsKg,
  });
}

class TripFuelMetricsResult {
  final double totalFuelLiters;
  final double idleFuelLiters;
  final double totalCost;
  final double idleCost;
  final double costSavedCruising;
  final double cruisingEfficiencyLPer100Km;

  TripFuelMetricsResult({
    required this.totalFuelLiters,
    required this.idleFuelLiters,
    required this.totalCost,
    required this.idleCost,
    required this.costSavedCruising,
    required this.cruisingEfficiencyLPer100Km,
  });
}

class FuelEngine {
  VehicleConfig config;

  FuelEngine([VehicleConfig? config])
      : config = config ?? defaultVehicleConfig;

  VehicleConfig updateConfig(VehicleConfig newConfig) {
    config = newConfig;
    return config;
  }

  VehicleConfig getConfig() {
    return config;
  }

  /// Calculate fuel wasted and monetary cost in LKR (Sri Lankan Rupees) during idle stationary state.
  IdleWasteResult calculateIdleWaste(int idleDurationSeconds) {
    final idleHours = idleDurationSeconds / 3600;
    final idleLiters = idleHours * config.idleConsumptionRateLph;
    final idleCost = idleLiters * config.fuelPricePerLiter;
    
    // Cost per minute while stuck idling in LKR
    final costPerMinute = (config.idleConsumptionRateLph / 60) * config.fuelPricePerLiter;

    // Approx CO2 emissions per Liter of gasoline = ~2.31 kg CO2
    final co2EmissionsKg = idleLiters * 2.31;

    return IdleWasteResult(
      idleLiters: double.parse(idleLiters.toStringAsFixed(3)),
      idleCost: double.parse(idleCost.toStringAsFixed(2)),
      costPerMinute: double.parse(costPerMinute.toStringAsFixed(2)),
      co2EmissionsKg: double.parse(co2EmissionsKg.toStringAsFixed(2)),
    );
  }

  /// Calculate overall trip fuel cost breakdown in LKR.
  TripFuelMetricsResult calculateTripFuelMetrics(
      double distanceKm, int idleDurationSeconds, double avgSpeedKmH) {
    final idleWaste = calculateIdleWaste(idleDurationSeconds);
    final idleLiters = idleWaste.idleLiters;
    final idleCost = idleWaste.idleCost;

    // Cruising efficiency model (approx 6.5 - 9.0 L/100km depending on engine size)
    final baseLPer100Km = 5.5 + config.engineSizeLiters * 1.2;
    
    // Speed efficiency curve multiplier (optimal speed ~60-80 km/h)
    double speedPenalty = 1.0;
    if (avgSpeedKmH > 0 && avgSpeedKmH < 30) {
      speedPenalty = 1.35; // Stop-and-go traffic jam penalty
    } else if (avgSpeedKmH > 100) {
      speedPenalty = 1.25; // High speed drag penalty
    }

    final cruisingLPer100Km = baseLPer100Km * speedPenalty;
    final cruisingLiters = (distanceKm / 100) * cruisingLPer100Km;
    final totalFuelLiters = cruisingLiters + idleLiters;

    final cruisingCost = cruisingLiters * config.fuelPricePerLiter;
    final totalCost = cruisingCost + idleCost;

    // Hypothetical savings if 70% of idle traffic jam time was smooth cruising
    final hypotheticalSavedIdleLiters = idleLiters * 0.7;
    final costSavedCruising = hypotheticalSavedIdleLiters * config.fuelPricePerLiter;

    return TripFuelMetricsResult(
      totalFuelLiters: double.parse(totalFuelLiters.toStringAsFixed(2)),
      idleFuelLiters: idleLiters,
      totalCost: double.parse(totalCost.toStringAsFixed(2)),
      idleCost: idleCost,
      costSavedCruising: double.parse(costSavedCruising.toStringAsFixed(2)),
      cruisingEfficiencyLPer100Km: double.parse(cruisingLPer100Km.toStringAsFixed(1)),
    );
  }
}
